solver/llbsolver/file/backend.go

```diff
@@ -66,7 +68,11 @@ func mapUserToChowner(user *copy.User, idmap *idtools.IdentityMapping) (copy.Cho
 }
 
 func mkdir(ctx context.Context, d string, action pb.FileActionMkDir, user *copy.User, idmap *idtools.IdentityMapping) error {
-	p, err := fs.RootPath(d, filepath.Join("/", action.Path))
+	actionPath, err := system.NormalizePath("/", action.Path, runtime.GOOS, false)
```

Comment:
`runtime.GOOS` ?

Reply:
> So in effect, the path needs to be native to the host we're running on. 

Input or output path. The path coming from LLB should be unix and input path is what os param should control and `NormalizePath` says it always returns unix path as well.

Reply:
> Input or output path. The path coming from LLB should be unix and input path is what os param should control and `NormalizePath` says it always returns unix path as well.

In this context, slashes do not matter. A Windows host is perfectly capable of dealing with paths that have mixed slashes. The really important reason we call ```NormalizePath()``` with ```runtime.GOOS``` here, is to make sure that we don't end up with paths like: ```C:/ProgramData/containerd/root/C:/some/path```. ```NormalizePath()``` makes sure to strip away any drive letter and validates we don't have ```UNC``` paths, among other things. Not doing so will lead to an invalid path, but not because of slashes.

On the other hand, the following path is perfectly valid on linux and should not be mutated:

```
/var/lib/myStrangeApp/C:/IMadePoorChoicesWhenDefiningPathsOnLinux
```

The following path will work on Windows:

```
C:\\var/a\\asdsd.txt
```

Sample:

```go
package main

import (
	"fmt"
	"log"
	"os"
)

func main() {
	pth := "C:\\var/aa\\asdsd.txt"
	info, err := os.Stat(pth)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("%s is there\n", info.Name())
}
```

Which on my system yields:

```powershell
PS C:\Users\Administrator> go run test.go
asdsd.txt is there
```

Reply:
hmm. I think you're right. By the time this is called, the path should have already be normalized once. So this path should already be fine (drive letter stripped).

I think we can pass ```action.Path``` directly to ```fs.RootPath()``` without needing to go through either ```NormalizePath```` or ```filepath.Join()```.

Reply:
This is a variation of the discussion above about nil pointers. In this case ```system.NormalizePath()``` is idempotent. We can call it several times in a row and the result should be the same, if the path is correct. We normalize the path in the client before marshaling it and sending it over, but not everyone will use the ```go``` client to interact with buildkit.

We should either attempt to normalize the path here as well, or return an error if the expectation is that the path **must** be normalized when it reaches this point.

What are your thoughts? Either way is fine with me.